### ISSUE TITLE
Error "path already exists" on add new child

### ISSUE THREAD

**Issue by brabant** [django-treebeard/django-treebeard#81]
An error occurs when creating a new node in the CMS Wagtail "this path already exists." 
I change the line 277 in the file models.py and then work normaly
return self.get_children().order_by("path").reverse()[0]

**Comment by bkfox** [comment]
Hey, I also have got this error; there is also something I don't understand: is not "last()" method used in this case? This allows returning None when the node is a leaf.

**Comment by patta42** [comment]
The same happens here, although I cannot provide steps on how to reproduce this. Interestingly, even adding the `order_by('path')` to the return value of `MP_Node.get_children` does not help, but adding it to in the `get_last_child` method does. I suggest to add it in the `get_first_child` method, too.

Another interesting thing is that when I call the `get_last_child` method of the same node from the shell (opened via `manage.py shell`), the result is correct.

Here is my data:

Wagtail: 1.13.1
Django: 1.11.7
Python: 3.4.6

In my opinion, this is something which definitely requires more research. If this happens on a production server, one cannot add new pages to the corresponding tree, which is a severe problem.

**Comment by gasman** [comment]
Without any steps to reproduce, I don't think it's possible to say whether this is an issue with treebeard, Wagtail or the user's own project.

**Comment by patta42** [comment]
Uhhh, here's the problem, at least for my case. I was too blind to see it, sorry.

Okay, here are the steps to reproduce the error:

Make a wagtail `Page`-model which overrides the `get_children()` method and changes the order of the children.

```
class Testpage ( Page ):
    def get_children( self ):
        return super(Testpage, self).get_children().order_by('title')
```

1. Create a Testpage instance
2, Add a child with title `c` to the Testpage
3. Add a child with title `a` to the Testpage
4. Adding another child will now throw the error.


The reason is very obvious: the `get_last_child()` method of `Node` relies on the unchanged order of the result of its `get_children()` method, which was obviously changed by the model above.

I don't know whether this should be considered a bug or not, I would suggest to add a method `_get_children()` to `Node`, with the underscore indicating that it should not be touched, and then use this method in the `get_last_child()` method (and `get_first_child()` accordingly).

At least I would suggest to update the docs to state that `get_children()` should not be changed.

**Comment by gasman** [comment]
The problem with marking a method as 'do not override this' (via naming convention, or in the documentation) is that it sets an expectation that any methods that _aren't_ marked are fair game. And that would mean that the burden is on treebeard to anticipate any possible ways that a user might screw things up through method overriding - that doesn't really seem feasible to me :-/ Ultimately I don't think we can really escape the fact that method overriding is "do this at your own risk".

**Comment by patta42** [comment]
To be honest, I expect every method not marked as 'do not override' as safe to be changed, as long as it still does what the method name or docs say. In this particular case, the method still returns all children. I think that adding `order_by('path')` or introducing a `_get_children()` method would be easy. Of course, an additional, changeable, `get_children()` should be implemented, too:
```
def get_children(self):
    return self._get_children()
```

Note that the same applies to the  `get_next_sibling()` and `get_prev_sibling()` methods.

Alas, I'm not familiar with github, but I can try to update the code if you want.

**Comment by philippeluickx** [comment]
I have the same issue, just trying to test my application using treebeard. I do not use wagtail.
The issue is obvious when printing out the queryset:

```
Add Project1

Add Project2
get_last_child returns "Project 1" (correct)
<QuerySet [<Document: Project1>]>

Add Project3
get_last_child returns "Project 1" (not correct)
<QuerySet [<Document: Project1>, <Document: Project2>]>

```

The queryset is not sorted on path, which should be as the OP mentioned.

**Comment by jrief** [comment]
closed via #112
